src/SpendPermissionManager.sol

```diff
@@ -202,13 +250,77 @@ contract SpendPermissionManager is EIP712 {
         _transferFrom(spendPermission.token, spendPermission.account, spendPermission.spender, value);
     }
 
+    /// @notice Approve a spend permission batch via a signature from the account.
+    ///
+    /// @param spendPermissionBatch Details of the spend permission batch.
+    /// @param signature Signed approval from the user.
+    function approveBatchWithSignature(SpendPermissionBatch memory spendPermissionBatch, bytes calldata signature)
+        public
+    {
+        bytes32 batchHash = getBatchHash(spendPermissionBatch);
+        if (_isInvalidated[batchHash][spendPermissionBatch.account]) {
+            revert InvalidatedBatch();
+        }
+        // validate signature over spend permission batch data
+        if (
+            IERC1271(spendPermissionBatch.account).isValidSignature(batchHash, signature)
+                != IERC1271.isValidSignature.selector
+        ) {
+            revert InvalidSignature();
+        }
+
+        uint256 batchLen = spendPermissionBatch.tokenAllowances.length;
+        for (uint256 i; i < batchLen; i++) {
+            _approve(
+                SpendPermission({
+                    account: spendPermissionBatch.account,
+                    spender: spendPermissionBatch.tokenAllowances[i].spender,
+                    start: spendPermissionBatch.start,
+                    end: spendPermissionBatch.end,
+                    period: spendPermissionBatch.period,
+                    token: spendPermissionBatch.tokenAllowances[i].token,
+                    allowance: spendPermissionBatch.tokenAllowances[i].allowance,
+                    salt: spendPermissionBatch.tokenAllowances[i].salt,
+                    extraData: spendPermissionBatch.tokenAllowances[i].extraData
+                })
+            );
+        }
+    }
+
     /// @notice Hash a SpendPermission struct for signing in accordance with EIP-712.
     ///
     /// @param spendPermission Details of the spend permission.
     ///
     /// @return hash Hash of the spend permission.
     function getHash(SpendPermission memory spendPermission) public view returns (bytes32) {
-        return _hashTypedData(keccak256(abi.encode(MESSAGE_TYPEHASH, spendPermission)));
+        return _hashTypedData(keccak256(abi.encode(PERMISSION_TYPEHASH, spendPermission)));
+    }
+
+    /// @notice Hash a SpendPermissionBatch struct for signing in accordance with EIP-712.
+    ///
+    /// @param spendPermissionBatch Details of the spend permission batch.
+    ///
+    /// @return hash Hash of the spend permission batch.
+    function getBatchHash(SpendPermissionBatch memory spendPermissionBatch) public view returns (bytes32) {
```

Comment:
not sure if better here or in `approveBatchWithSignature`, but we could tighten our object rules with non-zero length of `tokenAllowances` and unique entries via enforced sorting. Can change next week though and take in chunks